packages/core/src/storage/didcomm/DidCommMessageRecord.ts

```diff
@@ -0,0 +1,99 @@
+import type { AgentMessage } from '../../agent/AgentMessage'
+import type { JsonMap } from '../../types'
+import type { DidCommMessageRole } from './DidCommMessageRole'
+
+import { AriesFrameworkError } from '../../error'
+import { JsonTransformer } from '../../utils/JsonTransformer'
+import { rightSplit } from '../../utils/string'
+import { isJsonMap } from '../../utils/type'
+import { uuid } from '../../utils/uuid'
+import { BaseRecord } from '../BaseRecord'
+
+export type DefaultDidCommMessageTags = {
+  role: DidCommMessageRole
+  associatedRecordId?: string
+
+  // Computed
+  protocolName: string
+  messageName: string
+  versionMajor: string
+  versionMinor: string
+  messageType: string
+  messageId: string
+  threadId: string
+}
+
+export interface DidCommMessageRecordProps {
+  role: DidCommMessageRole
+  message: JsonMap
+  id?: string
+  createdAt?: Date
+  associatedRecordId?: string
+}
+
+export class DidCommMessageRecord extends BaseRecord<DefaultDidCommMessageTags> {
+  public message!: JsonMap
+  public role!: DidCommMessageRole
+
+  /**
+   * The id of the record that is associated with this message record.
+   *
+   * E.g. if the connection record wants to store an invitation message
+   * the associatedRecordId will be the id of the connection record.
+   */
+  public associatedRecordId?: string
+
+  public static readonly type = 'DidCommMessageRecord'
+  public readonly type = DidCommMessageRecord.type
+
+  public constructor(props: DidCommMessageRecordProps) {
+    super()
+
+    if (props) {
+      this.id = props.id ?? uuid()
```

Comment:
I'm thinking if we shouldn't use message ID instead of creating an artificial one. But I guess it's actually a record ID and not a message ID in this case.

Reply:
Problem with message id is that it can be provided by other agents, meaning it could be non-unique.... That's why I went with a new, generated, id.